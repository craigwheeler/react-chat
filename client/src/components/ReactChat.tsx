import React, { useEffect, useState, useRef } from 'react';
import qs from 'qs';
import io from 'socket.io-client';
import styled from 'styled-components';

// server address for local dev is 'http://127.0.0.1:8000'
// no param passed to connect when deployed to heroku
const socket = io.connect();

const Messages = () => {
  // Get username and chatId from URL
  const { username, chatId } = qs.parse(window.location.search, {
    ignoreQueryPrefix: true
  });
  const [messages, setMessages] = useState<any>([]);
  const msgRef = useRef<HTMLDivElement>(document.createElement('div'));

  const renderMessages = () =>
    messages.map((item: any, key: any) => (
      <div className="message" key={key}>
        <p className="meta">
          {item.username}: <span> {item.time}</span>
        </p>
        <p ref={msgRef} className="text">
          {item.text}
        </p>
      </div>
    ));

  const scrollToBottom = () => {
    msgRef.current.scrollIntoView({ behavior: 'smooth' });
  };

  useEffect(() => {
    // Message from server
    socket.on('message', (msg: any) => {
      setMessages((message: any) => [...message, msg]);
    });
  }, []);

  useEffect(() => {
    // Join chatroom
    socket.emit('joinRoom', { username, chatId });
  }, [chatId, username]);

  useEffect(() => {
    if (!messages.length) return;
    scrollToBottom();
  }, [messages]);

  return (
    <main className="chat-main">
      <div className="chat-messages">{renderMessages()}</div>
    </main>
  );
};

const Input = () => {
  const onSubmit = (e: any) => {
    e.preventDefault();
    const msg = e.target.elements.msg.value.trim();
    if (!msg) return;
    // Emit message to server
    socket.emit('chatMessage', msg);
    // Clear input
    e.target.elements.msg.value = '';
    e.target.elements.msg.focus();
  };

  return (
    <div className="chat-form-container">
      <form id="chat-form" onSubmit={(e) => onSubmit(e)}>
        <input id="msg" type="text" placeholder="Type your message..." required />
        <button className="btn">Send</button>
      </form>
    </div>
  );
};

const ReactChat = () => {
  return (
    <ChatContainer>
      <Messages />
      <Input />
    </ChatContainer>
  );
};

export default ReactChat;

const ChatContainer = styled.div`
  background: #fff;
  overflow: hidden;
  .btn {
    cursor: pointer;
    padding: 5px 15px;
    background: var(--light-color);
    color: var(--dark-color-a);
    border: 0;
    font-size: 17px;
  }

  .chat-header {
    background: var(--dark-color-a);
    color: #fff;
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
    padding: 15px;
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  .chat-messages {
    padding: 30px;
    height: calc(100vh - 40px);
    overflow-y: scroll;
    .message {
      padding: 10px;
      margin-bottom: 15px;
      background-color: var(--light-color);
      border-radius: 5px;
      overflow-wrap: break-word;
      .meta {
        font-size: 12px;
        font-weight: bold;
        color: var(--dark-color-b);
        opacity: 0.7;
        margin-bottom: 7px;
        span {
          color: #777;
        }
      }
      .text {
        font-size: 14px;
      }
    }
  }

  .chat-form-container {
    border-top: 1px solid rgb(232, 232, 232);
    form {
      display: flex;
    }
    input[type='text'] {
      width: 100%;
      padding: 12px 14px;
      outline: 0;
      border: 0;
    }
  }
`;
