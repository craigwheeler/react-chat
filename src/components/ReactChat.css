@import url('https://fonts.googleapis.com/css?family=Roboto&display=swap');

* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

:root {
  --dark-color-a: #667aff;
  --dark-color-b: #7386ff;
  --dark-color-c: #222;
  --light-color: #e6e9ff;
  --success-color: #5cb85c;
  --error-color: #d9534f;
}

body {
  font-family: 'Roboto', sans-serif;
  font-size: 16px;
  background: var(--dark-color-c);
}

ul {
  list-style: none;
}

a {
  text-decoration: none;
}

.btn {
  cursor: pointer;
  padding: 5px 15px;
  background: var(--light-color);
  color: var(--dark-color-a);
  border: 0;
  font-size: 17px;
}

.chat-container {
  background: #fff;
  overflow: hidden;
}

.chat-header {
  background: var(--dark-color-a);
  color: #fff;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  padding: 15px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.chat-sidebar {
  background: var(--dark-color-b);
  color: #fff;
  padding: 20px 20px 60px;
  overflow-y: scroll;
}

.chat-sidebar h2 {
  font-size: 20px;
  background: rgba(0, 0, 0, 0.1);
  padding: 10px;
  margin-bottom: 20px;
}

.chat-sidebar h3 {
  margin-bottom: 15px;
}

.chat-sidebar ul li {
  padding: 10px 0;
}

.chat-messages {
  padding: 30px;
  height: calc(100vh - 40px);
  overflow-y: scroll;
}

.chat-messages .message {
  padding: 10px;
  margin-bottom: 15px;
  background-color: var(--light-color);
  border-radius: 5px;
  overflow-wrap: break-word;
}

.chat-messages .message .meta {
  font-size: 12px;
  font-weight: bold;
  color: var(--dark-color-b);
  opacity: 0.7;
  margin-bottom: 7px;
}

.chat-messages .message .text {
  font-size: 14px;
}

.chat-messages .message .meta span {
  color: #777;
}

.chat-form-container {
  border-top: 1px solid rgb(232, 232, 232);
}

.chat-form-container form {
  display: flex;
}

.chat-form-container input[type='text'] {
  width: 100%;
  padding: 12px 14px;
  outline: 0;
  border: 0;
}
