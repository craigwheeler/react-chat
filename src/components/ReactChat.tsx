import React from 'react';
import './ReactChat.css';

const ReactChat = () => {
  return (
    <div className="chat-container">
      <main className="chat-main">
        <div className="chat-messages"></div>
      </main>
      <div className="chat-form-container">
        <form id="chat-form">
          <input id="msg" type="text" placeholder="Type your message..." required />
          <button className="btn">Send</button>
        </form>
      </div>
    </div>
  );
};

export default ReactChat;
