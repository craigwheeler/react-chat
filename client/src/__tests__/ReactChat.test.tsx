import React from 'react';
import { render } from '@testing-library/react';
import ReactChat from '../components/ReactChat';

describe('React Chat', () => {
  it('renders without crashing', () => {
    const element_tree = render(<ReactChat />).container;
    expect(element_tree).toMatchSnapshot();
  });
});
