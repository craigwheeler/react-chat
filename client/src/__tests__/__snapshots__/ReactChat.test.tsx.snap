// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Chat renders without crashing 1`] = `
<div>
  <div
    class="sc-bdnylx jWOIGI"
  >
    <main
      class="chat-main"
    >
      <div
        class="chat-messages"
      />
    </main>
    <div
      class="chat-form-container"
    >
      <form
        id="chat-form"
      >
        <input
          id="msg"
          placeholder="Type your message..."
          required=""
          type="text"
        />
        <button
          class="btn"
        >
          Send
        </button>
      </form>
    </div>
  </div>
</div>
`;
